geopandas/tools/clip.py

```diff
@@ -0,0 +1,246 @@
+
+    # Clip the data with the polygon
+    if isinstance(gdf_sub, GeoDataFrame):
+        clipped = gdf_sub.copy()
+        clipped["geometry"] = gdf_sub.intersection(poly)
+
+        # Return the clipped layer with no null geometry values or empty geometries
+        return clipped[~clipped.geometry.is_empty & clipped.geometry.notnull()]
+    else:
+        # GeoSeries
+        clipped = gdf_sub.intersection(poly)
+        return clipped[~clipped.is_empty & clipped.notnull()]
+
+
+def clip(gdf, clip_obj, keep_geom_type=False):
+    """Clip points, lines, or polygon geometries to the clip_obj extent.
+
+    Both layers must be in the same Coordinate Reference System (CRS).
+    The `gdf` will be clipped to the full extent of the clip object.
+
+    If there are multiple polygons in clip_obj, data from `gdf` will be
+    clipped to the total boundary of all polygons in clip_obj.
+
+    Parameters
+    ----------
+    gdf : GeoDataFrame or GeoSeries
+        Vector layer (point, line, polygon) to be clipped to clip_obj.
+    clip_obj : GeoDataFrame, GeoSeries, (Multi)Polygon
+        Polygon vector layer used to clip `gdf`.
+        The clip_obj's geometry is dissolved into one geometric feature
+        and intersected with `gdf`.
+    keep_geom_type : boolean
+        If True, return only geometries of original type in case of intersection
+        resulting in multiple geometry types or GeometryCollections.
+        If False, return all resulting geometries (potentially mixed-types).
+
+    Returns
+    -------
+    GeoDataFrame or GeoSeries
+         Vector data (points, lines, polygons) from `gdf` clipped to
+         polygon boundary from clip_obj.
+
+    Examples
+    --------
+    Clip points (global cities) with a polygon (the South American continent):
+
+    >>> import geopandas
+    >>> path =
+    >>> world = geopandas.read_file(
+    ...     geopandas.datasets.get_path('naturalearth_lowres'))
+    >>> south_america = world[world['continent'] == "South America"]
+    >>> capitals = geopandas.read_file(
+    ...     geopandas.datasets.get_path('naturalearth_cities'))
+    >>> capitals.shape
+    (202, 2)
+    >>> sa_capitals = geopandas.clip(capitals, south_america)
+    >>> sa_capitals.shape
+    (12, 2)
+    """
+    if not isinstance(gdf, (GeoDataFrame, GeoSeries)):
+        raise TypeError(
+            "'gdf' should be GeoDataFrame or GeoSeries, got {}".format(type(gdf))
+        )
+
+    if not isinstance(clip_obj, (GeoDataFrame, GeoSeries, Polygon, MultiPolygon)):
+        raise TypeError(
+            "'clip_obj' should be GeoDataFrame, GeoSeries or"
+            "(Multi)Polygon, got {}".format(type(gdf))
+        )
+
+    if isinstance(clip_obj, (GeoDataFrame, GeoSeries)):
+        xmin, ymin, xmax, ymax = clip_obj.total_bounds
+    else:
+        xmin, ymin, xmax, ymax = clip_obj.bounds
+    if gdf.cx[xmin:xmax, ymin:ymax].empty:
+        raise ValueError("gdf and clip_obj extent do not overlap.")
```

Comment:
If it's empty, shouldn't the clip return a NullGeometry/NaN, in the same manner as `sjoin`?

Reply:
> if it's empty, shouldn't the clip return a NullGeometry/NaN, in the same manner as sjoin?

I  suppose you mean an empty GeoDataFrame? (now rows, but the column names can be preserved)

But indeed, that might be more consistent

Reply:
Good point. I'll change that.